consensus/common/src/main/java/tech/pegasys/pantheon/consensus/common/jsonrpc/AbstractGetSignerMetricsMethod.java

```diff
@@ -0,0 +1,127 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.LongStream;
+
+public abstract class AbstractGetSignerMetricsMethod {
+
+  private final BlockchainQueries blockchainQueries;
+  private final VoteTallyCache voteTallyCache;
+  private final JsonRpcParameter parameters;
+
+  public AbstractGetSignerMetricsMethod(
+      final BlockchainQueries blockchainQueries,
+      final VoteTallyCache voteTallyCache,
+      final JsonRpcParameter parameter) {
+    this.blockchainQueries = blockchainQueries;
+    this.voteTallyCache = voteTallyCache;
+    this.parameters = parameter;
+  }
+
+  public JsonRpcResponse response(final JsonRpcRequest request) {
+
+    final long startBlock = parameters.required(request.getParams(), 0, Long.class);
+    final long endBlock = parameters.required(request.getParams(), 1, Long.class);
+
+    if (isValidParameters(startBlock, endBlock)) {
+
+      final Map<Address, ProposerReportBlockProductionResult> proposersMap = new HashMap<>();
+      final Map<Address, ValidatorReportBlockProductionResult> validatorInLastBlockMap =
+          new HashMap<>();
+
+      // go through each block (startBlock is inclusive and endBlock is exclusive)
+      LongStream.range(startBlock, endBlock)
+          .boxed()
+          .sorted(Collections.reverseOrder())
+          .forEach(
+              currentIndex -> {
+                Optional<BlockHeader> blockHeaderByNumber =
+                    blockchainQueries.getBlockHeaderByNumber(currentIndex);
+
+                blockHeaderByNumber.ifPresent(
+                    header -> {
+                      if (validatorInLastBlockMap.isEmpty()) {
+                        // Get all validators present in the last block of the range
+                        voteTallyCache
+                            .getVoteTallyAfterBlock(header)
+                            .getValidators()
+                            .forEach(
+                                address ->
+                                    validatorInLastBlockMap.put(
+                                        address,
+                                        new ValidatorReportBlockProductionResult(address)));
+                      }
+
+                      // Get the number of blocks from each proposer in a given block range.
+                      getProposersOfBlock(header)
+                          .forEach(
+                              proposerAddress -> {
+                                final ProposerReportBlockProductionResult proposer =
+                                    proposersMap.get(proposerAddress);
+                                if (proposer != null) {
+                                  proposer.incrementeNbBlock();
+                                } else {
+                                  proposersMap.put(
+                                      proposerAddress,
+                                      new ProposerReportBlockProductionResult(proposerAddress));
```

Comment:
The `Map.computeIfAbsent` is really helpful here.  If `ProposerReportBlockProductionResult` defaulted to 0 blocks proposed instead of 1, then you could just use `proposersMap.computeIfAbsent(proposerAddress, ProposerReportBlockProductionResult::new).incrementeNbBlock();`